lib/lifecycle.js

```diff
@@ -0,0 +1,728 @@
+// Copyright (c) 2020 Wayne Parrott. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+const rclnodejs = require('bindings')('rclnodejs');
+const LifecyclePublisher = require('./lifecycle_publisher.js');
+const loader = require('./interface_loader.js');
+const Service = require('./service.js');
+
+const SHUTDOWN_TRANSITION_LABEL = rclnodejs.getLifecycleShutdownTransitionLabel();
+
+// An instance of State message constructor used for accessing State 
+// state machine constants. This interface is lazy initialized at runtime.
+let StateInterface; 
+
+// An instance of Transition message constructor used for accessing Transition 
+// state machine constants. This interface is lazy initialized at runtime.
+let TransitionInterface;
+
+function getStateInterface() {
+  if (!StateInterface) {
+    StateInterface = loader.loadInterface('lifecycle_msgs/msg/State');
+  }
+  return StateInterface;
+}
+
+function getTransitionInterface() {
+  if (!TransitionInterface) {
+    TransitionInterface = loader.loadInterface('lifecycle_msgs/msg/Transition');
+  }
+  return TransitionInterface;
+}
+
+/**
+ * @typedef SerializedState
+ * @type {object}
+ * @property {number} id - code identifying the type of this state.
+ * @property {string} label - readable name of this state.
+ */
+
+/**
+ * The state of the lifecycle state model.
+ */
+class State {
+
+  /**
+   * Create a state.
+   * @param {number} id - The id value.
+   * @param {string} label - The label value.
+   */
+  constructor(id, label) {
+    this.id = id;
+    this.label = label;
+  }
+
+  /**
+   * Create an object representation of state properties.
+   * @return {SerializedState} The object.
+   */
+  asMessage() {
```

Comment:
Doesn't this return the same thing except it removes the prototype chain?

Reply:
+1. Removed asMessage() methods from State, Transition and TransitionDescription.